rfd 226: Stranded Connection Recovery

dboslee opened the pull request:
Open up discussion around improving behavior around failure and recovery. This focuses on the issue of agents getting stranded connected to unhealthy proxy and auth services and proposes solutions to each of these.


Rendered Markdown:
https://github.com/gravitational/teleport/blob/rfd/226-self-healing/rfd/0226-stranded-connection-recovery.md

dboslee committed:
rfd: availability, failure and self-healing

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+
+### Desired State of Failure Recovery
+
+Teleport should automatically detect when control plane components become unhealthy
+and recover without manual intervention. Agents should be able to seamlessly
```
I'm focused on agents here but if these components use a gRPC client they could also take advantage of the custom load balance policy. I would need more context on their API usage or other types of connections they are creating to comment on whether they would benefit from the proposed changes here or propose additional changes.

dboslee replied:
This is a grpc dial option which should already be configurable for the api.Client via the config here https://github.com/gravitational/teleport/blob/c15f090a06142c3ee8a3235094a3b1c9e5679757/api/client/client.go/#L621-L622

The only thing we need to do is specify the `grpc.WithDefaultServiceConfig(serviceConfig)` for the clients we want to have the new behavior. Clients which do not specify this option will continue behaving the same. The service config will look similar to what we specify for access graph here https://github.com/gravitational/teleport.e/blob/c580e73a0a6b5fe5f00b2e8f8af0647846b0b015/lib/accessgraph/accessgraph.go/#L50-L55

accept the name of the load balancer policy will be the custom load balancer policy we implement and register rather than `round_robin`.

I will update the doc accordingly.

dboslee replied:
updated in [53add5f](https://github.com/gravitational/teleport/pull/59159/commits/53add5f60efa938969be9cc4f4cc7f6787549c8b)

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+
+### Current State of Failure Recovery
+
+When a Teleport control plane component (Auth or Proxy) becomes unhealthy, the
+agents connected to that component may become unavailable. Recovery in this
```
If the relay service uses the reversetunnel agent library it would also benefit from these changes. But I will need more context on how that service is implemented.

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+not automatically reconnect in this scenario. Within gRPC client connections are
+maintained by load balancing policies[^2].
+
+Evaluating the default `pick_first` policy. It will only create a new connection
+when the existing TCP connection is disrupted. When health checking is enabled
+the client begins failing requests immediately rather than sending them to the
+server but never attempts to reconnect.
```
In the go-grpc library you can enable health checking on the pick_first lb policy. But to clarify, this section is describing behavior I observed during testing and explaining why the `pick_first` policy does not work for our use case.

Further down I describe a custom LB policy implementation I am proposing we use which will support health checks.

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+Proxy. The solution should be indifferent to the transport.
+
+We can communicate the health of the Auth instance an agent is connected to using
+gRPC health checking[^1]. This is a standardized service API that gRPC servers can implement. Clients that enable health checking can then poll or stream the health
```
👍 good point, I did not specify which we would use. I figured we would implement both but I believe clients, by default, will prefer to use the streaming API when its available.

Lets assume thats the case for now. If it turns out clients still use the unary call or require additional configuration to use the streaming API we will adjust as necessary to ensure the streaming API is used.

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+in response to an agent heartbeat[^8] to avoid sending the signal to all connected
+agents at the same time.
+
+Currently, agents stop accepting new transport requests while draining[^9]. This
+could impact agent availability if we begin to drain and an agent is unable to
+establish a new connection to the cluster. To resolve this issue, after receiving a
```
> If the service orchestrator at the proxy end has detected that a proxy is unhealthy and has spawned a new one, wouldn't it also just terminate the unhealthy one

In my experience restarting a service is reserved for the most extreme failures where the service is completely unresponsive. Like Teleport's `/healthz` endpoint timing out. As an operator I would assume an unhealthy service can recover on its own but I should stop sending client traffic to it in the mean time.

> How can we be sure that there's distinct proxies available for reconnection?

This feature also assumes that traffic will be rerouted to healthy proxy services. I think we should have this be opt-in behavior either via config or env var. This should help avoid it being used incorrectly.

> aren't we just going to keep slamming the remaining proxies with new handshakes from the entire chunk of the cluster that was connected through the proxy that became unhealthy?

How else would we solve this without reconnecting? We should certainly have backoff/jitter when retyring connections to avoid slamming the proxies though.

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+while leaving old connections open for long lived streams to continue.
+
+#### Periodic Tunnel Reconnects
+With the improved agent draining behavior described in [Proxy Reconnects](#proxy-reconnects)
+it should be safe to introduce periodic reversetunnel reconnects. This would
+allow for better load distribution and the most up-to-date low latency routing paths
+for these connections. There is some risk that long lived sessions could keep draining
+tunnels open for a very long period of time. Given enough of these types of sessions
+the total number of open tunnels would slowly climb and potentially cause issues.
```
I think the customer is requesting that agents periodically reconnect regardless of proxy mode. So its not directly related to the agent isolation issue we are trying to solve here but it could be solved by the same mechanism. However it sounds like the two should be enabled/configured separately.

I'll just assume env vars for now so something like

```
# Configure teleport proxy server to send reconnect to agents after being connected for duration + jitter
TELEPORT_REVERSETUNNEL_PERIODIC_RECONNECT_DURATION="24h"
# Configure teleport proxy server to send reconnect after being unhealthy for duration.
TELEPORT_REVERSETUNNEL_UNHEALTHY_RECONNECT_DURATION="1m"
```

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+in response to an agent heartbeat[^8] to avoid sending the signal to all connected
+agents at the same time.
+
+Currently, agents stop accepting new transport requests while draining[^9]. This
+could impact agent availability if we begin to drain and an agent is unable to
+establish a new connection to the cluster. To resolve this issue, after receiving a
+reconnect request, agents will first establish a new connection to a healthy proxy 
+instance. Only then will the agent begin to drain the connection and stop accepting
+new transport requests.
```
To the time it takes to close the tunnel. The tunnel should only be terminated once it has no transports. Allowing new transports while we're trying to shut it down could increase the time it takes to shutdown.

dboslee committed:
explicitly mention use of streaming health check api

dboslee committed:
update link to pick first policy

mend

dboslee committed:
add details on using custom lb policy

dboslee committed:
update tunnel sections

dboslee committed:
update auth reconnect sections

dboslee committed:
update section on heartbeat interval configuration

espadolini commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,246 @@
+
+The new behavior will only send Proxy discovery requests when there is a Proxy
+heartbeat and will only send the set of proxies that have heartbeated since the 
+last Proxy discovery request was sent.
```
This is equivalent to saying "will only send the new proxy heartbeat".

espadolini commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+in network traffic. However given gRPC's graceful shutdown process this may not
+be an issue. Assuming the streams are not a major contributor to the typical auth
+load we see it may be beneficial to periodically move new requests to new connections
+while leaving old connections open for long lived streams to continue.
```
We can't resume a `WatchEvents` stream in general and adding the ability to do so for the backends for which it could be possible would require a decent amount of effort.

espadolini reviewed (commented):
The health checking and channel switching for grpc doesn't fill me with confidence - how do we know that there is a new auth ready to take requests, without interacting with the load balancer itself? Aren't we potentially just multiplying the connections and increasing the load on the same auth that's already not doing well?

What about the few auth APIs that are still used over HTTP? We could just disable the keepalive on the HTTP client and call it a day, but if that's the strategy we should mention it in the doc.

dboslee committed:
remove pointer

dboslee committed:
add info on policy config compatibility

dboslee committed:
rename expiry duration to ttl

dboslee commented:
@espadolini regarding

>What about the few auth APIs that are still used over HTTP? We could just disable the keepalive on the HTTP client and call it a day, but if that's the strategy we should mention it in the doc.

I think setting `DisableKeepAlives` is a good fallback plan. But as a slightly more efficient option what do you think about a [roundtripper](https://pkg.go.dev/net/http#RoundTripper) which sets [response.Close = true](https://pkg.go.dev/net/http#Response:~:text=//%20Close%20records%20whether%20the%20header%20directed%20that%20the%20connection%20be%0A%09//%20closed%20after%20reading%20Body.%20The%20value%20is%20advice%20for%20clients%3A%20neither%0A%09//%20ReadResponse%20nor%20Response.Write%20ever%20closes%20a%20connection.%0A%09Close%20bool) when there is a `5xx` or `4xx` error code. That way connections can still be reused until there is some client/server error.

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,179 @@
+shutdown and another new connection will be created.
+
+If the old connection recovers before a new connection is able to reach the `READY`
+state, we will stop creating new connections and begin using the old connection again.
```
To keep the behavior similar to today we can continue sending requests to the old unhealthy connection until there is a new healthy connection to send requests to. This avoids failing requests that could succeed due to auth reporting unhealthy but it only being partially degrade (for example KMS is failing but backend ops are still able to succeed).

dboslee replied:
I've updated the doc to describe the graceful behavior.

dboslee committed:
expand gRPC lb policy behavior and add notes

dboslee committed:
fix formatting

dboslee committed:
clarify heartbeat behavior

dboslee committed:
update discovery request description

dustinspecker commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,286 @@
+
+Any open RPCs on the old connection will be allowed to run to completetion.
+
+Clients will discover the new load balancing policy via the `/webapi/ping` endpoint.
```
I was talking with @sclevine and he suggested using `/webapi/find` over `/webapi/ping` since find is a cached result, while ping hits auth each time.

dustinspecker commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,286 @@
+made until either the existing connection returns to `READY` or the new connection
+reaches the `READY` state.
+
+All connection attempts should use a backoff policy with jitter to avoid a thundering
```
Curious how we want to handle retrieving the service configuration as well. Is jitter enough to prevent issues with all agents making HTTP requests against Proxy to see if any changes to service configuration have occurred?

Also curious what Teleport agents do today if a Proxy shuts down. Not sure if a new client connection is created or how that reconnect logic works. Would we be safe to only query the find endpoint once during connect/reconnect? As of now, a Proxy server would have to restart to begin returning a different service configuration. Can we avoid polling altogether?

dboslee committed:
lb config discovery via grpc and forceful stream termination

dboslee committed:
add section on http clients

dboslee committed:
env var moved to auth with grpc lb policy discovery

dboslee committed:
update http client section

dboslee committed:
describe reasoning behind defering forceful rpc closing

dboslee committed:
rename rfd and focus on connections

originally there were some loftier goals here but we're reducing
scope and updating the title to reflect that

dboslee committed:
add some headers for organization of large sections

dboslee committed:
fix typos

rosstimothy commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,382 @@
+fields could be added in the future.
+
+This can be configured on an auth server by setting the environment variable
+`TELEPORT_UNSTABLE_GRPC_CLIENT_LB_POLICY`. When not specified clients will continue
```
Does this need to be set on the Proxy as well? Will the Proxy always include `grpc_client_lb_policy` regardless of this setting?

rosstimothy commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,382 @@
+
+### Proxy Reconnects
+
+#### Background on reversetunnels
```
@okraport how do the changes that you are currently working on to make reverse tunnel clients aware of networking issues effect this proposal?

okraport commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,382 @@
+
+#### Closing Connections
+
+Today the agent never closes reversetunnel connections. This can lead to more reversetunnel connections open than expected when proxies become unhealthy and recover. 
```
@dboslee I have been looking at this code in the context of detecting stale TCP connections when the default interface is changed/reconfigured.  Which would be a similar condition to the proxy being unresponsive from the agent PoV.  

In the current situation the agent writes pings to the heartbeat channel without waiting for reply, if the link is stale the messages are just buffered and the connection is dropped after TCP is dead which could be as much as 3min by default. 

My idea here would be to make the server respect WantReply on heartbeats so that the agent could theoretically detect whether the connection is healthy and allow us to drop it quicker than the TCP timeout. Of course this has a risk of dropping connections too eagerly, one thing @espadolini has suggested is to only do this when no traffic has been received from the proxy side for {some-period}. Then try to get a response. I would like to hear your thoughts on this since you have spent some time investigating this area.

dboslee committed:
clarify use of grpc/protobuf and add protobufs

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,382 @@
+called immediately after connecting and the initial health checking is complete.
+
+As a backup plan clients could also discover the new load balancing policy via the
+`/webapi/ping` endpoint. However the gRPC endpoint is the preferred solution.
```
updated to clarify we will pursue grpc only.

dboslee committed:
add some clarification around ttl field

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,246 @@
+
+The `ExpiryDuration` for each server will be set based on the heartbeat interval
+configured at each proxy. This will be stored in the backend `Server` resource
+as `Server.Spec.TTL`. This allows different values to exist on different services
```
I've clarified its a new field. Let me know if there is any specific questions on the reasoning behind adding this.

This field will be propagated down to agents so they can base their tracker expiry off of the TTL. I had worries about using expiry for this purpose so I am proposing the new field. The difference is expiry is a time and ttl is a duration.

dboslee committed:
be very explicit on server ttl field

dboslee committed:
specify how config updates are handled

dboslee committed:
remove server ttl field

dboslee commented on rfd/0226-self-healing.md:
```diff
@@ -0,0 +1,458 @@
+The `TTL` used in discovery requests will be set based on the a value stored in
+the backend `Server`. This field `Server.Spec.TTL` will be added to store this value.
+This new field is a duration set based on the `TELEPORT_UNSTABLE_PROXY_ANNOUNCE_TTL`
+or the default `ServerAnnounceTTL`.
```
I've decided to just drop the field and use the local proxies configured ttl instead. I also highlight the potential downsides to this approach.

[4fc240b](https://github.com/gravitational/teleport/pull/59159/commits/4fc240ba95c77ce8ccebaca4199ac8265a64f3e6)

dboslee committed:
Update rfd/0226-self-healing.md

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

dboslee committed:
update custom lb policy

dboslee committed:
fix mermaid sequence

dboslee committed:
rename file to  match title

dboslee committed:
add brief description of diagram

dboslee committed:
fix typos

okraport commented on rfd/0226-stranded-connection-recovery.md:
```diff
@@ -0,0 +1,520 @@
+3. What if the old connection becomes healthy again? If the old connection becomes
+healthy before we we're able to establish a new healthy connection we will stop
+attempting new connections and continue using that old connection. Otherwise it
+will we shutdown.
```
```suggestion
will be shutdown.
```

okraport reviewed (commented):
Only minor typos

dboslee committed:
Update rfd/0226-stranded-connection-recovery.md

Co-authored-by: Luke Okraszewski <lukasz.okraszewski@goteleport.com>

dboslee committed:
Update rfd/0226-stranded-connection-recovery.md

Co-authored-by: Luke Okraszewski <lukasz.okraszewski@goteleport.com>

dboslee committed:
fix lint
